core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -901,6 +900,67 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
+  def handleCreateTopicRequest(request: RequestChannel.Request) {
+    val createTopicRequest = request.body.asInstanceOf[CreateTopicRequest]
+
+    def handleUnauthorizedRequest = {
+      createTopicRequest.topics.asScala.map { case (topic, _) =>
+        (topic, Errors.CLUSTER_AUTHORIZATION_FAILED)
+      }
+    }
+
+    def handleAuthorizedRequest = {
+      val brokers = metadataCache.getAliveBrokers.map(_.id)
+      createTopicRequest.topics.asScala.map { case (topic, arguments) =>
+        try {
+          // Prep configs
+          val configs = new Properties()
+          arguments.configs.asScala.foreach { case (key, value) =>
+            configs.setProperty(key, value)
+          }
+          LogConfig.validate(configs)
+
+          // Prep assignment
+          val assignments = {
+            if (!arguments.replicasAssignments.isEmpty) {
+              // Note: we don't check that replicaAssignment doesn't contain unknown brokers - unlike in add-partitions case,
+              // this follows the existing logic in TopicCommand
+              arguments.replicasAssignments.asScala.map { case (partitionId, replicas) =>
+                (partitionId.intValue, replicas.asScala.map(_.intValue).toSeq)
+              }
+            } else {
+              AdminUtils.assignReplicasToBrokers(brokers, arguments.partitions, arguments.replicationFactor)
+            }
+          }
+
+          // Create the topic
+          AdminUtils.createOrUpdateTopicPartitionAssignmentPathInZK(zkUtils, topic, assignments, configs, update = false)
+
+          // Return valid response
+          (topic, Errors.NONE)
+        } catch {
+          // Handle errors with appropriate error response
+          case e: Throwable => (topic, Errors.forException(e))
+        }
+      }
+    }
+
+    val results =
+      if(!authorize(request.session, Create, Resource.ClusterResource)) {
+        handleUnauthorizedRequest
```

Comment:
This is definitely a style thing...this may be a bit too long of a response but I don't want to just say "because I like it".

This class (KafkaApis) is growing very large. Independent of adding these new handlers I would really like to break out each handler (or group of related handlers into their own class/file to help readability/clarity/etc. I didn't want to do that here because KIP-4 is critical functionality and I wanted to keep the change/impact minimal to get it reviewed and committed more quickly.

If I were writing this functionality in its own class file I would follow "good code design" by breaking up low level functionality into expressive methods and compose them for the handling of the protocol. Things along the lines of "methods should do one thing" and "methods should descend 1 level of abstraction". I'm no purist, but breaking things out can help find reusable code pieces, provide clarity for higher level methods, simplify refactoring/evolution and help show missing test coverage (that actually matters).

This choice of using inline definitions is a half-way/half-backed way of doing it without the full refactor I would like and may not be perfect, but I thought it separated two branches of logic clearly. Authorized requests and unauthorized requests.